tools/approval-service/internal/approvalservice/githubevents/githubevents.go

```diff
@@ -0,0 +1,231 @@
+package githubevents
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/google/go-github/v69/github"
+	"github.com/gravitational/shared-workflows/libs/github/webhook"
+	"github.com/gravitational/shared-workflows/tools/approval-service/internal/approvalservice/config"
+)
+
+// Source is a webhook that listens for GitHub events and processes them.
+type Source struct {
+	processor   DeploymentReviewEventProcessor
+	addr        string
+	secretToken string
+
+	deployReviewChan chan *github.DeploymentReviewEvent
+	listener         net.Listener
+	srv              *http.Server
+
+	validation map[string]struct{}
+
+	log *slog.Logger
+}
+
+// DeploymentReviewEventProcessor is an interface for processing deployment review events.
+type DeploymentReviewEventProcessor interface {
+	// ProcessDeploymentReviewEvent processes a deployment review event.
+	// Automated checks are done before this is called.
+	// If the automated checks fail, valid will be false.
+	ProcessDeploymentReviewEvent(event DeploymentReviewEvent, valid bool) error
+}
+
+// DeploymentReviewEvent is an event that is sent when a deployment review is requested.
+// This contains information needed to process a request
+// If multiple underlying events/payloads/etc. roll under
+// the same "root" event that approval is for, they should
+// all set the same ID.
+type DeploymentReviewEvent struct {
+	Requester    string
+	Environment  string
+	Organization string
+	Repository   string
+	WorkflowID   int64
+	// Other fields TODO. Potential fields:
+	// * Source
+	// * Commit/tag/source control identifier
+	// * "Parameters" map. In the case of GHA, this would be
+	//   any input provided to a workflow dispatch event
+	// See RFD for more details
+}
+
+// Opt is a functional option for the GitHub event source.
+type Opt func(*Source)
+
+// WithLogger sets the logger for the GitHub event source.
+func WithLogger(logger *slog.Logger) Opt {
+	return func(s *Source) {
+		s.log = logger
+	}
+}
+
+var defaultOpts = []Opt{
+	WithLogger(slog.Default()),
+}
+
+func NewSource(cfg config.GitHubEvents, processor DeploymentReviewEventProcessor, opt ...Opt) *Source {
+	s := &Source{
```

Comment:
nit: consistent naming with receiver funcs
```suggestion
	ghes := &Source{
```